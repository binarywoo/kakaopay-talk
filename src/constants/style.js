export const COLOR = {
  PRIMARY: '#2d1c05',
  YELLOW: '#ffdc00'
}
