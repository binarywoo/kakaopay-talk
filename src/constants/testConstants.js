export const DUMMY_IMG_URL =
  'https://s.pstatic.net/static/www/img/2018/sp_search.svg'

export const DUMMY_CHAT = {
  title: 'test chat',
  lastUpdate: '2019-04-06T18:32:11.433Z',
  lastMessage: 'test message',
  key: 'testKey'
}

export const DUMMY_MESSAGE = {
  content: '하이',
  key: '-LbnOJfNoldq3SD0xRIR',
  lastUpdate: '2019-04-06T16:04:55.757Z',
  type: 'text',
  user: '-LbmTnQw-OjNS87fDbBt',
  userId: '이웅희'
}

export const DUMMY_IMAGE_MESSAGE = {
  content: '하이',
  key: '-LbnOJfNolds3SD0xRIR',
  lastUpdate: '2019-04-06T16:06:55.757Z',
  type: 'image',
  user: '-LbmTnQw-OjNS87fDbBt',
  userId: '이웅희'
}

export const DUMMY_USER = {
  key: '-LbmTnQw-OjNS87fDbBt',
  userId: '이웅희'
}
