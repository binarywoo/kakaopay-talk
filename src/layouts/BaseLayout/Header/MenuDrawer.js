import React, { useCallback, useState } from 'react'
import { Drawer, Button, message } from 'antd'
import PictureWall from '../../../components/PictureWall/PictureWall'
import { GithubPicker } from 'react-color'

const MenuDrawer = ({
  show,
  user,
  onChangeBackgroundColor,
  onClose,
  form,
  onApplyProfileImage
}) => {
  const [profileImage, setProfileImage] = useState(null)
  const [isLoading, setIsLoading] = useState(false)

  const getProfileInfoFromUser = useCallback(() => {
    if (user && user.profileImage) {
      const splitFileInfo = user.profileImage.split('/')
      const userProfileImage = [
        {
          uid: -1,
          name: splitFileInfo[splitFileInfo.length - 1],
          status: 'done',
          url: user.profileImage
        }
      ]
      return userProfileImage
    } else {
      return null
    }
  }, [user])
  const normFile = useCallback(e => {
    if (Array.isArray(e)) return e
    return e && e.fileList
  }, [])

  const onClickApply = useCallback(() => {
    form.validateFields(err => {
      if (!err) {
        if (
          profileImage &&
          profileImage.fileList &&
          profileImage.fileList.length > 0
        ) {
          setIsLoading(true)
          const profileFileObj = profileImage.fileList[0].originFileObj
          onApplyProfileImage(profileFileObj).then(() => {
            message.success('저장되었습니다.')
            setIsLoading(false)
          })
        }
      }
    })
  }, [profileImage])

  const onChangeProfileImage = useCallback(e => {
    setProfileImage(e)
  }, [])

  return (
    <Drawer
      placement='right'
      closable={true}
      visible={show}
      onClose={onClose}
      className='menu-drawer'
      title={
        <div>
          <span className='fw-7000'>{user ? user.userId : null}</span>님 로그인
          중
        </div>
      }
    >
      <div className='setting-zone'>
        <h4>프로필 사진</h4>
        {form.getFieldDecorator('profileImage', {
          valuePropName: 'fileList',
          getValueFromEvent: normFile,
          initialValue: getProfileInfoFromUser()
        })(<PictureWall onChange={onChangeProfileImage} max={1} useCrop />)}
        <Button
          onClick={onClickApply}
          loading={isLoading}
          disabled={
            !profileImage ||
            !profileImage.fileList ||
            profileImage.fileList.length === 0
          }
          style={{ marginBottom: 24 }}
        >
          프로필 사진 저장
        </Button>
        <h4>채팅방 배경색</h4>
        <GithubPicker
          colors={[
            '#B80000',
            '#DB3E00',
            '#FCCB00',
            '#008B02',
            '#006B76',
            '#1273DE',
            '#004DCF',
            '#5300EB'
          ]}
          triangle='hide'
          onChange={onChangeBackgroundColor}
        />
      </div>
    </Drawer>
  )
}

export default MenuDrawer
