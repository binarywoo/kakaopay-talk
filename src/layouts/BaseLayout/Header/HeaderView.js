import React, { useCallback, useEffect } from 'react'
import { PageHeader, Affix, notification, Button, Modal, Icon } from 'antd'
import { convertSnapshotToArr } from '../../../libs/daoUtils'
import { USER_KEY } from '../../../constants/commonConstants'

const HeaderView = ({
  header,
  history,
  headerAction,
  invitations,
  firebase,
  user,
  subscribeAction,
  invitationService,
  userAction
}) => {
  useEffect(() => {
    // user의 초대 내역을 구독
    if (user) {
      firebase
        .database()
        .ref(`invitations/${user.key}`)
        .on('value', snapshot => {
          subscribeAction.receiveSubscribe(
            'invitations',
            convertSnapshotToArr(snapshot)
          )
        })
    }
  }, [])

  const onAcceptInvitation = useCallback(
    invitation => {
      notification.destroy(invitation.key)
      invitationService.putInvitation(
        `${invitation.to}/${invitation.key}/checked`,
        true
      )
      history.push(`/chat/${invitation.chat}`)
    },
    [invitations]
  )

  const onDenyInvitation = useCallback(invitation => {
    notification.destroy(invitation.key)
    invitationService.putInvitation(
      `${invitation.to}/${invitation.key}/checked`,
      true
    )
  })

  const onClickLogout = useCallback(() => {
    Modal.confirm({
      title: '로그아웃',
      content: '정말 로그아웃 하시겠습니까?',
      okText: '예',
      cancelText: '아니오',
      onOk: () => {
        userAction.clearUser()
        localStorage.removeItem(USER_KEY)
      }
    })
  }, [user])

  useEffect(() => {
    const newInvitations = invitations
      ? invitations.filter(invitation => !invitation.checked)
      : null
    if (newInvitations && newInvitations.length > 0) {
      newInvitations.forEach(newInvitation => {
        notification.open({
          duration: 0,
          message: '새로운 초대 도착',
          description: `${newInvitation.fromId}님으로 부터 ${
            newInvitation.chatTitle
          }방에 초대되었습니다.`,
          key: newInvitation.key,
          btn: [
            <Button
              key='deny'
              onClick={() => onDenyInvitation(newInvitation)}
              style={{ marginRight: 5 }}
            >
              거절
            </Button>,
            <Button
              key='accept'
              type='primary'
              onClick={() => onAcceptInvitation(newInvitation)}
            >
              수락
            </Button>
          ]
        })
      })
    }
  }, [invitations])

  const onBack = useCallback(() => {
    if (header.backPath) {
      return () => {
        history.push(header.backPath)
        headerAction.receiveBackPath(null)
      }
    } else {
      return null
    }
  }, [header])

  return (
    <Affix offsetTop={0}>
      <PageHeader
        className='kakao-header'
        onBack={onBack()}
        title={header.title || '까까오톡'}
        extra={
          header.extra ||
          (user && (
            <div style={{ cursor: 'pointer' }} onClick={onClickLogout}>
              <Icon type='logout' />
            </div>
          ))
        }
      />
    </Affix>
  )
}

export default HeaderView
