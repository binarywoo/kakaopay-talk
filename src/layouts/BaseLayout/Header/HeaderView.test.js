import React from 'react'
import { shallow } from 'enzyme'
import HeaderView from './HeaderView'
import { fromJS } from 'immutable'
import { Icon, Affix, PageHeader } from 'antd'
import { DUMMY_USER } from '../../../constants/testConstants'
import MenuDrawer from './MenuDrawer'

describe('HeaderView', () => {
  let wrapper = null

  const props = {
    header: {
      title: 'test',
      extra: <Icon type='export' />
    },
    invitations: [],
    user: DUMMY_USER,
    form: {},
    onAcceptInvitation: jest.fn(),
    onDeynyInvitation: jest.fn(),
    onLogout: jest.fn(),
    onBack: jest.fn(),
    onChangeBackgroundColor: jest.fn(),
    onApplyProfileImage: jest.fn()
  }

  it('정상적으로 렌더링이 된다.', () => {
    wrapper = shallow(<HeaderView {...props} />)
  })

  it('스냅샷과 일치한다.', () => {
    expect(wrapper).toMatchSnapshot()
  })

  it('하위 컴포넌트들을 정상적으로 렌더링한다.', () => {
    expect(wrapper.find(Affix).length).toBe(1)
    expect(wrapper.find(PageHeader).length).toBe(1)
    expect(wrapper.find(MenuDrawer).length).toBe(1)
  })

  it('추가 컴포넌트를 정상적으로 렌더링한다.', () => {
    expect(wrapper.find(PageHeader).prop('extra').length).toBe(3)
    let newProps = fromJS(props)
      .setIn(['header', 'extra'], null)
      .toJS()
    wrapper = shallow(<HeaderView {...newProps} />)
    expect(wrapper.find(PageHeader).prop('extra').length).toBe(2)
    newProps = Object.assign({}, newProps, {
      user: null
    })
    wrapper = shallow(<HeaderView {...newProps} />)
    expect(wrapper.find(PageHeader).prop('extra').length).toBe(0)
  })

  it('타이틀을 정상적으로 렌더링한다.', () => {
    expect(wrapper.find(PageHeader).prop('title')).toBe('test')
    const newProps = fromJS(props)
      .setIn(['header', 'title'], null)
      .toJS()
    wrapper = shallow(<HeaderView {...newProps} />)
    expect(wrapper.find(PageHeader).prop('title')).toBe('까까오톡')
  })
})
