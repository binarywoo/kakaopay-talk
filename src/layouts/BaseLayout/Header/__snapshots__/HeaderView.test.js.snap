// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HeaderView 스냅샷과 일치한다. 1`] = `
<Affix
  offsetTop={0}
  target={[Function]}
>
  <PageHeader
    backIcon={
      <Icon
        type="arrow-left"
      />
    }
    className="kakao-header"
    extra={
      Array [
        <Icon
          type="export"
        />,
        <div
          onClick={[Function]}
          style={
            Object {
              "cursor": "pointer",
              "display": "inline-block",
              "marginRight": 24,
            }
          }
        >
          <Icon
            type="logout"
          />
        </div>,
        <div
          onClick={[Function]}
          style={
            Object {
              "cursor": "pointer",
              "display": "inline-block",
            }
          }
        >
          <Icon
            type="setting"
          />
        </div>,
      ]
    }
    title="test"
  />
  <MenuDrawer
    form={Object {}}
    onApplyProfileImage={[MockFunction]}
    onChangeBackgroundColor={[MockFunction]}
    onClose={[Function]}
    show={false}
    user={
      Object {
        "key": "-LbmTnQw-OjNS87fDbBt",
        "userId": "이웅희",
      }
    }
  />
</Affix>
`;
